### DISCUSSION TITLE
Issues with building boost dependency in Windows 11

### DISCUSSION THREAD

**Discussion by loorthu** [AcademySoftwareFoundation/OpenRV#470] [Build Issues]
I am trying to build openrv on a Windows 11 system and getting stuck where the boost build fails. The build abruptly ends with this cryptic error:

```
     1>Done Building Project "C:\OpenRV\_build\cmake\dependencies\RV_DEPS_BOOST.vcxproj" (default targets) -- FAILED.

Build FAILED.

       "C:\OpenRV\_build\cmake\dependencies\RV_DEPS_BOOST.vcxproj" (default target) (1) ->
       (CustomBuild target) ->
         C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom build for 'C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-build.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-install.rule;C:\OpenRV\_build\CMakeFiles\ddfe0fbca828d26d63401d1d79fd8a3b\RV_DEPS_BOOST-complete.rule;C:\OpenRV\_build\CMakeFiles\dc0b82080efe119cccb8b58ef9885864\RV_DEPS_BOOST.rule' exited with code 1. [C:\OpenRV\_build\cmake\dependencies\RV_DEPS_BOOST.vcxproj]

    0 Warning(s)
    1 Error(s)
```

I am attaching the complete log of `rvbuildt RV_DEPS_BOOST` for reference. I tried the fixes mentioned in https://github.com/AcademySoftwareFoundation/OpenRV/discussions/458 but no luck so far. 

Appreciate any guidance.. thanks!

[openrv_build_boost.log](https://github.com/AcademySoftwareFoundation/OpenRV/files/15438308/openrv_build_boost.log)

**Comment by bernie-laberge** [comment]
Hello @loorthu !

I apologize for the delay in getting back to you.

I just rebuilt from scratch the latest Open RV [commit](https://github.com/AcademySoftwareFoundation/OpenRV/commit/0ba6a733689557cc42b925c4bbefcae02fa9d07d) on my Windows system and it built without errors. (Typical it works on my system answer sorry!).

When you get the chance, could you please confirm that you still have this build issue ?

I will use a brand new windows machine and reinstall everything from scratch. Sometimes MSYS2 gets updated and it can cause issues with the Open RV build. You might be experiencing that...

**Comment by loorthu** [comment]
Hi @bernie-laberge 

Thank you for the response. We have tried different permutations (Windows 11 vs Windows 10, VS 2022 vs VS 2019 etc) and have ended up in different dead-ends with dependency build issues.

One related issue could be that the latest build instructions refer to using VS 2022 but the screen shots in https://github.com/AcademySoftwareFoundation/OpenRV/blob/main/docs/build_system/win_microsoft_visual_studio_features_to_install.png refer to VS 2019.

Is it possible to give us a verified build environment with:

- Windows edition used
- Visual Studio version
- Visual Studio installation details about exactly what SDKs are needed (similar to the screenshot mentioned above) for the above VS version
- The openrv commit tag or branch that is used (e.g v2.0.1)
- Expected console output while running "rvcfg" and "rvbuild" for the above environment. This will let us diff the expected one with the actual one and compare where things differ

This way, we can spin off a vanilla Windows VM with the above specifications and see if we are able to build it.

Thanks!

**Reply by bernie-laberge** [reply]
These are the components on my Windows 11 system:

- OS Name	Microsoft Windows 11 Enterprise, Version	10.0.22000 Build 22000
- Visual Studio Professional 2022 17.7.6
- Open RV commit : 0ba6a733689557cc42b925c4bbefcae02fa9d07d

**Comment by bernie-laberge** [comment]
<img width="1245" alt="visual_studio_options" src="https://github.com/AcademySoftwareFoundation/OpenRV/assets/117092886/356cc586-bc69-4e57-855d-a76f0eadd6bc">

**Comment by loorthu** [comment]
Hi,

My system specs are below:

- Windows 11 Pro, Version 23H2, OS build 22631.3593 (x64-based processor)
- Visual Studio 2022 Community Edition (with the same python and c++ options chosen as yours)
- Using latest mingw64 shell for building environment
- Openrv commit 0ba6a733689557cc42b925c4bbefcae02fa9d07d

The build fails with the following message at the end:

```
       "C:\OpenRV\_build\main_executable.vcxproj" (default target) (1) ->
       "C:\OpenRV\_build\src\bin\apps\rv\rv.vcxproj" (default target) (3) ->
       "C:\OpenRV\_build\cmake\dependencies\RV_DEPS_BOOST.vcxproj" (default target) (5) ->
       (CustomBuild target) -> 
         C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom build for 'C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-mkdir.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-download.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-update.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-patch.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-configure.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-build.rule;C:\OpenRV\_build\CMakeFiles\7f794420ad2eee1f7b2f7e3966c7c970\RV_DEPS_BOOST-install.rule;C:\OpenRV\_build\CMakeFiles\ddfe0fbca828d26d63401d1d79fd8a3b\RV_DEPS_BOOST-complete.rule;C:\OpenRV\_build\CMakeFiles\dc0b82080efe119cccb8b58ef9885864\RV_DEPS_BOOST.rule;C:\OpenRV\cmake\dependencies\CMakeLists.txt' exited with code 1. [C:\OpenRV\_build\cmake\dependencies\RV_DEPS_BOOST.vcxproj]

    4941 Warning(s)
    1 Error(s)
```

I am also attaching the console output of both rvcfg and rvbuild for your reference.

Do you think the Professional vs Community version of Visual Studio might be making a difference?

[openrv_build.log](https://github.com/user-attachments/files/15568856/openrv_build.log)
[openrv_cfg.log](https://github.com/user-attachments/files/15568857/openrv_cfg.log)

**Reply by bernie-laberge** [reply]
Hello @loorthu,
It shouldn't make a difference whether you use the Community or the Professional Visual Studio.
Here are my own logs:
[bernie_rvbuild.log.zip](https://github.com/user-attachments/files/15588220/bernie_rvbuild.log.zip)

**Reply by bernie-laberge** [reply]
Note that you can re-launch the build of a specific target using the following command:
`rvbuildt RV_DEPS_BOOST`
(There is an additional 't' added to the rvbuild command which is an alias to build a specific target)

**Reply by bernie-laberge** [reply]
When I compare my build log against your build log, I notice that you have an additional message. I wonder if it is relevant:
Your log:
```    5>PrepareForBuild:
         Creating directory "x64\Release\RV_DEPS_BOOST\".
         Structured output is enabled. The formatting of compiler diagnostics will reflect the error hierarchy. See https://aka.ms/cpp/structured-output for more details.
         Creating directory "x64\Release\RV_DEPS_BOOST\RV_DEPS_BOOST.tlog\".
```

My log:
```     6>PrepareForBuild:
         Creating directory "x64\Release\RV_DEPS_BOOST\".
         Creating directory "x64\Release\RV_DEPS_BOOST\RV_DEPS_BOOST.tlog\".
```

I don't have this 'Structured output is enabled.' Strange...

**Reply by bernie-laberge** [reply]
After looking it up I don't think it is relevant to the issue at hand.

**Reply by bernie-laberge** [reply]
There seems to be something strange with how the boost build exe b2.exe (which is not part of the Open RV build per actually part of the Boost build) is detecting your hardware:

The boost-b2.exe detected architecture of your build machine is

default architecture : none [1]
Mine is
default architecture : x86 [1]
Is there something peculiar with your build machine by chance ?


```
Yours:
     5>CustomBuild:
         Performing configuration checks
         
             - default address-model    : none [1]
             - default architecture     : none [1]
             - compiler supports SSE2   : no [2]
             - compiler supports SSE4.1 : no [2]
             - has synchronization.lib  : no [2]
             - has std::atomic_ref      : no [2]
             - has statx                : no [2]
             - has statx syscall        : no [2]
             - has BCrypt API           : no [2]
             - is Windows CE            : no [2]
             - has init_priority attribute : no [2]
             - has stat::st_blksize     : no [2]
             - has stat::st_mtim        : no [2]
             - has stat::st_mtimensec   : no [2]
             - has stat::st_mtimespec   : no [2]
             - has stat::st_birthtim    : no [2]
             - has stat::st_birthtimensec : no [2]
             - has stat::st_birthtimespec : no [2]
             - has fdopendir(O_NOFOLLOW) : no [2]
             - has POSIX *at APIs       : no [2]
             - has_icu builds           : no [2]
             - zlib                     : no
             - bzip2                    : no
             - lzma                     : no
             - zstd                     : no
             - has_lzma_cputhreads builds : no [2]
             - iconv (libc)             : no [2]
             - iconv (separate)         : no [2]
             - icu                      : no [2]
             - icu (lib64)              : no [2]
         
         [1] msvc-14.3
         [2] msvc-14.3/release/address-model-64/architecture-x86/cxxstd-17-iso/python-3.11/threadapi-win32/threading-multi/visibility-hidden
         
         Component configuration:
         
             - atomic                   : building
             - chrono                   : building
             - container                : not building
             - context                  : not building
             - contract                 : not building
             - coroutine                : not building
             - date_time                : building
             - exception                : not building
             - fiber                    : not building
             - filesystem               : building
             - graph                    : building
             - graph_parallel           : not building
             - headers                  : not building
             - iostreams                : building
             - json                     : not building
             - locale                   : building
             - log                      : not building
             - math                     : not building
             - mpi                      : not building
             - nowide                   : not building
             - program_options          : building
             - python                   : not building
             - random                   : building
             - regex                    : building
             - serialization            : building
             - stacktrace               : not building
             - system                   : building
             - test                     : not building
             - thread                   : building
             - timer                    : building
             - type_erasure             : not building
             - wave                     : not building


Mine:
         Performing configuration checks
         
             - default address-model    : 64-bit [1]
             - default architecture     : x86 [1]
             - compiler supports SSE2   : yes [2]
             - compiler supports SSE4.1 : yes [2]
             - has synchronization.lib  : yes [2]
             - has std::atomic_ref      : no [2]
             - has statx                : no [2]
             - has statx syscall        : no [2]
             - has BCrypt API           : yes [2]
             - has init_priority attribute : no [2]
             - has stat::st_blksize     : no [2]
             - has stat::st_mtim        : no [2]
             - has stat::st_mtimensec   : no [2]
             - has stat::st_mtimespec   : no [2]
             - has stat::st_birthtim    : no [2]
             - has stat::st_birthtimensec : no [2]
             - has stat::st_birthtimespec : no [2]
             - has fdopendir(O_NOFOLLOW) : no [2]
             - has POSIX *at APIs       : no [2]
             - has_icu builds           : no [2]
             - zlib                     : no
             - bzip2                    : no
             - lzma                     : no
             - zstd                     : no
             - has_lzma_cputhreads builds : no [2]
             - iconv (libc)             : no [2]
             - iconv (separate)         : no [2]
             - icu                      : no [2]
             - icu (lib64)              : no [2]
         
         [1] msvc-14.3
         [2] msvc-14.3/release/cxxstd-17-iso/python-3.11/threadapi-win32/threading-multi/visibility-hidden
         
         Component configuration:
         
             - atomic                   : building
             - chrono                   : building
             - container                : not building
             - context                  : not building
             - contract                 : not building
             - coroutine                : not building
             - date_time                : building
             - exception                : not building
             - fiber                    : not building
             - filesystem               : building
             - graph                    : building
             - graph_parallel           : not building
             - headers                  : not building
             - iostreams                : building
             - json                     : not building
             - locale                   : building
             - log                      : not building
             - math                     : not building
             - mpi                      : not building
             - nowide                   : not building
             - program_options          : building
             - python                   : not building
             - random                   : building
             - regex                    : building
             - serialization            : building
             - stacktrace               : not building
             - system                   : building
             - test                     : not building
             - thread                   : building
             - timer                    : building
             - type_erasure             : not building
             - wave                     : not building
```
Perhaps this could help ?
https://www.boost.org/build/tutorial.html

**Reply by loorthu** [reply]
Hi @bernie-laberge 

Thank you very much looking into the logs and finding relevant clues! It does sound like there is some issue in building boost, probably related to the specific laptop hardware I was using to build. Maybe a laptop vs desktop differences in hardware/windows features available to them. We will try this on a different windows system and see if we have better luck there.

**Reply by mpkepic** [reply]
https://github.com/microsoft/vcpkg/issues/38980

Note the patch to `boost-modular-build-helper`; 

For some reason the latest toolchain seems to be reporting as 144 which is throwing a bunch of things off.

**Reply by loorthu** [reply]
Thanks for the lead! The rv build scripts comes with the dependency to an older version of vcpkg and when I try to have it point to a newer version, it doesn't seem to like it. I tried the following in cmake/dependencies/vcpkg.cmake:

```
-  RV_CREATE_STANDARD_DEPS_VARIABLES("RV_DEPS_VCPKG" "2023.11.20" "" "")
+  RV_CREATE_STANDARD_DEPS_VARIABLES("RV_DEPS_VCPKG" "2024.04.23" "" "")
```

But that breaks the vckpg build with the following message:

```
  Performing download step (git clone) for 'RV_DEPS_VCPKG'
  Cloning into 'src'...
  fatal: invalid reference: 2024.04.23
  CMake Error at C:/OpenRV.boost_patch/_build/cmake/dependencies/RV_DEPS_VCPKG-prefix/tmp/RV_DEPS_VCPKG-gitclone.cmake:49 (message):
    Failed to checkout tag: '2024.04.23'
```

@bernie-laberge is it possible to see if you are able to reproduce this issue on Visual Studio 17.10? I noticed that your version is 17.7 and probably why it's not affected. Also let me know if my attempt at using a newer vcpkg is the right approach or if there is a better way.

**Reply by bernie-laberge** [reply]
Ah ! You might be right !
OK I will give it a try tomorrow.
I can easily create a new cloud Windows system and try the build instructions from scratch.
I will make sure the same version of Visual Studio as yours.
I will keep you posted.

**Reply by CraigMason** [reply]
Just to add a +1, I'm trying to build OpenRV today on a fresh Windows 11 VM and following all of the build instructions. I'm getting the same error and also have the same "none" architecture showing:

```
             - default address-model    : none [1]
             - default architecture     : none [1]
```

**Reply by bernie-laberge** [reply]
FYI: I was able to reproduce the Boost 1.80 build break !
@loorthu  was right : It is caused by the latest Visual Studio Community 2022 version 17.10.1 on a brand new Windows 11 system.
I had Visual Studio 2022 version 17.7.6 before and Boost 1.80 was happy with it. 
I am currently working on a fix.
I'll keep you posted.

**Reply by bernie-laberge** [reply]
Status Report:

There is an open issue on the Boost project for this Visual Studio 2022 17.10 compiler issue:
https://github.com/boostorg/boost/issues/914
This issue is still unresolved.

From that Boost issue thread, the following work around is proposed:
Install MSVC v141 - VS 2017 C++ x64/x86 build tools (v14.16)

I tested the work around and I was able to build Boost successfully in Open RV. 
I am now testing Open RV from scratch to make sure that everything else builds OK.

Here is the work around in pictures:

Step #1: Open Windows 'Add Remove Programs' and select to modify the installed version of Visual Studio:
<img width="889" alt="step1_add_remove_programs_modify_visual_studio" src="https://github.com/AcademySoftwareFoundation/OpenRV/assets/117092886/82b838fc-59e3-47fc-8297-21409f10e4aa">

Step #2: Add the option: MSVC v141 - VS 2017 C++ x64/x86 build tools (v14.16)
<img width="1237" alt="step2_add_option_MSVC_v141_VS_2017" src="https://github.com/AcademySoftwareFoundation/OpenRV/assets/117092886/4bdb06c6-0065-4769-8725-cb625f2b5bf9">

**Reply by CraigMason** [reply]
Hi @bernie-laberge 

Thanks for this workaround. I can confirm adding MSVC v141 as described allowed me to complete a full build and RV is now running on my Windows machine :)

Thanks for taking the time to look into this.

**Reply by loorthu** [reply]
Very happy to share that I was finally able to build OpenRV using the steps above! Thank you @bernie-laberge
